x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,350 @@
+
+    void start(long followGlobalCheckpoint) {
+        this.lastRequestedSeqno = followGlobalCheckpoint;
+        this.processedGlobalCheckpoint = followGlobalCheckpoint;
+        this.globalCheckpoint = followGlobalCheckpoint;
+
+        // Forcefully updates follower mapping, this gets us the leader imd version and
+        // makes sure that leader and follower mapping are identical.
+        updateMapping(imdVersion -> {
+            currentIndexMetadataVersion = imdVersion;
+            LOGGER.info("{} Started to follow leader shard {}, followGlobalCheckPoint={}",
+                params.getFollowShardId(), params.getLeaderShardId(), followGlobalCheckpoint);
+            coordinateReads();
+        });
+    }
+
+    private synchronized void coordinateReads() {
+        if (isStopped()) {
+            LOGGER.info("{} shard follow task has been stopped", params.getFollowShardId());
+            return;
+        }
+
+        LOGGER.trace("{} coordinate reads, lastRequestedSeqno={}, globalCheckpoint={}",
+            params.getFollowShardId(), lastRequestedSeqno, globalCheckpoint);
+        final long maxReadSize = params.getMaxReadSize();
+        while (hasReadBudget() && lastRequestedSeqno < globalCheckpoint) {
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            long size;
+            if (from + maxReadSize <= globalCheckpoint) {
+                size = maxReadSize;
+            } else {
+                size = globalCheckpoint - from;
+            }
+            LOGGER.trace("{}[{}] read [{}/{}]", params.getFollowShardId(), numConcurrentReads, from, size);
+            sendShardChangesRequest(from, size, from + size);
+            lastRequestedSeqno = from + size;
+        }
+
+        if (numConcurrentReads == 0) {
+            // We sneak peek if there is any thing new in the leader primary.
+            // If there is we will happily accept
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            LOGGER.trace("{}[{}] peek read [{}]", params.getFollowShardId(), numConcurrentReads, from);
+            sendShardChangesRequest(from, maxReadSize, null);
+        }
+    }
+
+    private boolean hasReadBudget() {
+        assert Thread.holdsLock(this);
+        if (numConcurrentReads >= params.getMaxConcurrentReads()) {
+            LOGGER.trace("{} no new reads, maximum number of concurrent reads have been reached [{}]",
+                params.getFollowShardId(), numConcurrentReads);
+            return false;
+        }
+        if (buffer.size() > params.getMaxBufferSize()) {
+            LOGGER.trace("{} no new reads, buffer limit has been reached [{}]", params.getFollowShardId(), buffer.size());
+            return false;
+        }
+        return true;
+    }
+
+    private synchronized void coordinateWrites() {
+        while (hasWriteBudget() && buffer.isEmpty() == false) {
+            if (numConcurrentWrites >= params.getMaxConcurrentWrites()) {
```

Comment:
This duplicates the `hasWriteBudget` check and can never fail by virtue of being inside the `while` loop.

Reply:
oops, I forgot to remove this.